rfcs/0148-pipe-operator.md

```diff
@@ -0,0 +1,327 @@
+but it is unknown and overlooked by many because it is not easily discoverable:
+Despite its great usefulness, it is currently used in less than 30 files in Nixpkgs
+(`rg '[\. ]pipe .* \['`).
+Additionally, it is not accessible to Nix code outside of nixpkgs,
+and due to Nix's lazy evaluation debugging type errors is really difficult.
+
+Let's have a look at an arbitrarily chosen snippet of Nixpkgs code:
+
+```nix
+defaultPrefsFile = pkgs.writeText "nixos-default-prefs.js" (lib.concatStringsSep "\n" (lib.mapAttrsToList (key: value: ''
+  // ${value.reason}
+  pref("${key}", ${builtins.toJSON value.value});
+'') defaultPrefs));
+```
+
+It is arguably pretty hard to read and reason about. Even when applying some more whitespace-generous formatting:
+
+```nix
+defaultPrefsFile = pkgs.writeText "nixos-default-prefs.js" (
+  lib.concatStringsSep "\n" (
+    lib.mapAttrsToList
+    (
+      key: value: ''
+        // ${value.reason}
+        pref("${key}", ${builtins.toJSON value.value});
+      ''
+    )
+    defaultPrefs
+  )
+);
+```
+
+One can observe the following issues:
+
+- If you want to follow the data flow, you must read it from bottom to top,
+  from the inside to the outside (the input here is `defaultPrefs`).
+- Adding a function call to the output would require wrapping the entire
+  expression in parentheses and increasing its indentation.
+
+Compare this to the equivalent call with `lib.pipe`:
+
+```nix
+defaultPrefsFile = pipe defaultPrefs [
+  (lib.mapAttrsToList (
+    key: value: ''
+      // ${value.reason}
+      pref("${key}", ${builtins.toJSON value.value});
+    ''
+  ))
+  (lib.concatStringsSep "\n")
+  (pkgs.writeText "nixos-default-prefs.js")
+];
+```
+
+The code now clearly reads from top to bottom in the order the data is processed,
+it is easy to add and remove processing steps at any point.
+
+With a dedicated pipe operator, it would look like this:
+
+```nix
+defaultPrefsFile = defaultPrefs
+  |> lib.mapAttrsToList (
+    key: value: ''
+      // ${value.reason}
+      pref("${key}", ${builtins.toJSON value.value});
+    ''
+  )
+  |> lib.concatStringsSep "\n"
+  |> pkgs.writeText "nixos-default-prefs.js";
+```
+
+The artificial distinction between the first input and the functions via the list now is gone,
+and so are the parentheses around the functions.
+With the lower character overhead, using the operator becomes attractive in more situations,
+whereas a `pipe` pays for its overhead only in more complex scenarios (usually three functions or more).
+Having a dedicated operator also increases visibility and discoverability of the feature.
+
+# Detailed design
+[design]: #detailed-design
+
+## `|>` operator
+
+A new operator `|>` is introduced into the Nix language.
+It is defined as function application with the order of arguments swapped: `f a` = `a |> f`.
+It is left-associative and has a binding strength one weaker than function application:
+`a |> f |> g b |> h` = `h ((g b) (f a))`.
```

Comment:
Explain this also in terms of `|>` + parenthesis - given that I genuinely understood nothing here

Indeed, as far as I remember, functions in Nix have only one argument (things like `a:b: a+b` are indeed a function receiving a function).
So, I'm confused about `h` here.

Reply:
I find this line pretty straightforward.

Not sure if this helps, but imagine the variables here have the following types:
```haskell
a :: a
b :: b
f :: a -> t1
g :: b -> t1 -> t2
h :: t2 -> t3
```

So each element in the pipeline after the first is a function taking the previous type and returning the next one:
```haskell
f :: a -> t1
g b :: t1 -> t2
h :: t2 -> t3
```

Reply:
Shepherds currently believe this conversation is resolved, outcome: RFC text does not need to change.